### PR TITLE
dynamic_modules: adds initial object loading logic

### PR DESCRIPTION
**Pull request by mathetake**
Commit Message: dynamic_modules: adds initial object loading logic
Additional Description:
This is the very first commit of the dynamic loading feature discussed among community
members. This is the effort to upstream the playground repository 
 https://github.com/mathetake/envoy-dynamic-modules as an Envoy core extension.
Series of commits will follow this little by little.

#2053, #24230, #32502

Risk Level: N/A (not compiled into the final build yet)
Testing: unit
Docs Changes: N/A
Release Notes: N/A 
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue] 
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/dynamic_modules/dynamic_modules.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+namespace DynamicModules {
+
+/**
+ * A class for loading and managing dynamic modules. This corresponds to a single dlopen handle.
+ * When the DynamicModule object is destroyed, the dlopen handle is closed.
+ *
+ * This class is supposed to be initialized once in the main thread and can be shared with other
+ * threads.
+ */
+class DynamicModule {
```
Comment (mathetake) [inline source/extensions/dynamic_modules/dynamic_modules.h:19]: note: this should be utilized in contrib/golang, and relatively easy, but  then the changes in this patch end up across many places, so I chose not to do that right now. - I could take a shot after this PR.

**Comment by mathetake** [comment]
I hope the test passes for non-linux, but meanwhile I will just wait and see 🤞

File: test/extensions/dynamic_modules/test_data/test_data.bzl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+# This declares a cc_library target that is used to build a shared library.
+# name + ".c" is the source file that is compiled to create the shared library.
+def test_program(name):
+    cc_name = name
+    cc_library(
+        name = cc_name,
```
Comment (mathetake) [inline test/extensions/dynamic_modules/test_data/test_data.bzl:8]: noticed unnecessary variable `cc_name` - will follow up later after i get reviews

File: source/extensions/dynamic_modules/dynamic_modules.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   * @param symbol_ref the symbol to look up.
+   * @return the symbol if found, otherwise nullptr.
+   */
+  void* getSymbol(const absl::string_view symbol_ref) const;
+
+  // The raw dlopen handle that can be used to look up symbols.
+  void* handle_ = nullptr;
+};
+
+using DynamicModulesSharedPtr = std::shared_ptr<DynamicModule>;
```
Comment (mathetake) [inline source/extensions/dynamic_modules/dynamic_modules.h:46]:
```suggestion
using DynamicModuleSharedPtr = std::shared_ptr<DynamicModule>;
```

File: CODEOWNERS [hunk 4]
```diff
@@ -345,8 +346,8 @@ extensions/filters/http/oauth2 @derekargueta @mattklein123
 /*/extensions/filters/http/match_delegate @wbpcode @jstraceski @tyxia
 # Generic proxy and related extensions
 /*/extensions/filters/network/generic_proxy/ @wbpcode @soulxu
-
-/*/extensions/health_checkers/common @zuercher @botengyao
```
Comment (mathetake) [inline CODEOWNERS:349]: moved this up as it looked like it was misplaced

**Comment by mathetake** [comment]
cc @mattklein123 @marc-barry

**Review by marc-barry — COMMENTED** [review]
This looks great and as it stands doesn't pose a risk (yet) as it isn't compiled into the final Envoy build at this time. 👍 to the approach of adding this little by little and starting with the initial object loading logic as the foundation for future PRs which will include code from https://github.com/mathetake/envoy-dynamic-modules. I really like the approach and simplicity of the ABI as defined in https://github.com/mathetake/envoy-dynamic-modules/tree/main/abi.

I didn't want to approve without this being reviewed by @mattklein123 who is the one assigned to this pull request.

File: source/extensions/dynamic_modules/dynamic_modules.h (hunk 1 above)
Comment (mattklein123) [reply source/extensions/dynamic_modules/dynamic_modules.h:20]: Yes please do, thanks.

File: source/extensions/dynamic_modules/dynamic_modules.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * threads.
+ */
+class DynamicModule {
+public:
+  DynamicModule(void* handle) : handle_(handle) {}
+  ~DynamicModule();
+
+  /**
+   * Get a function pointer from the dynamic module with a specific type.
+   * @tparam T the function pointer type to cast the symbol to.
```
Comment (mattklein123) [inline source/extensions/dynamic_modules/dynamic_modules.h:27]:
```suggestion
   * @param T the function pointer type to cast the symbol to.
```

File: source/extensions/dynamic_modules/dynamic_modules.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+private:
+  /**
+   * Get a symbol from the dynamic module.
+   * @param symbol_ref the symbol to look up.
+   * @return the symbol if found, otherwise nullptr.
+   */
+  void* getSymbol(const absl::string_view symbol_ref) const;
+
+  // The raw dlopen handle that can be used to look up symbols.
+  void* handle_ = nullptr;
```
Comment (mattklein123) [inline source/extensions/dynamic_modules/dynamic_modules.h:47]:
```suggestion
  void* handle_;
```

Always initialized in the constructor.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small nits.

/wait

**Comment by mathetake** [comment]
(CI is flaky as usual for unrelated reasons)

**Review by mattklein123 — APPROVED** [review]

File: test/extensions/dynamic_modules/dynamic_modules_test.cc [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  return TestEnvironment::substitute(
+             "{{ test_rundir }}/test/extensions/dynamic_modules/test_data/") +
+         "lib" + name + ".so";
+}
+
+TEST(DynamicModuleTest, InvalidPath) {
+  absl::StatusOr<DynamicModuleSharedPtr> result = newDynamicModule("invalid_name", false);
+  EXPECT_FALSE(result.ok());
+  EXPECT_EQ(result.status().code(), absl::StatusCode::kInvalidArgument);
+}
```
Comment (cpakulski) [inline test/extensions/dynamic_modules/dynamic_modules_test.cc:27]: Just drive by review: you should add a test which loads a file which exists, but is not a valid library.

**Pull request merged**